src/miniscript/evaluate.rs

```diff
@@ -0,0 +1,560 @@
+//! # Interpreatation of Miniscript scripts
+//!
+//! Traits and implementations to support for interpretation of miniscripts.
+//!
+//!
+use bitcoin_hashes::sha256;
+use secp256k1;
+
+use Error;
+use miniscript::astelem::AstElem;
+use ToPublicKey;
+use secp256k1::{Secp256k1, All, Signature};
+use bitcoin_hashes::Hash;
+
+///Primitives for Miniscript: a vector of public keys(sigs), hashlocks and timelocks
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct Primitives{
+    pub pks: Vec<bitcoin::PublicKey>,
+    pub hashlocks : Vec<Vec<u8>>,
+    pub timelocks : Vec<u32>
+}
+
+impl Primitives
+{
+    fn join(&mut self, other: &Primitives){
+        self.pks.extend(other.pks.clone());
+        self.hashlocks.extend(other.hashlocks.clone());
+        self.timelocks.extend(other.timelocks.clone());
+    }
+}
+
+/// Trait describing an AST element which can be evaluated /interpreted , given an input witness and
+/// maps from the public data to corresponding witness data.
+pub trait Interpretable<P> {
+    /// Attempt to interpret a miniscript for a given witness and
+    fn interpret(
+        &self,
+        witness: &mut Vec<Vec<u8>>,
+        age: u32,
+        msg: &secp256k1::Message,
+        secp: &Secp256k1<All>) -> Result<Primitives, Error>
+    where P : ToPublicKey + Clone;
+}
+
+impl <P: ToPublicKey + Clone> Interpretable<P> for AstElem<P> {
+    fn interpret(
+        &self,
+        witness: &mut Vec <Vec<u8>>,
+        age: u32,
+        msg: &secp256k1::Message,
+        secp: &Secp256k1<All>)
+        -> Result<Primitives, Error>
+    {
+        match *self {
+            AstElem::Pk(ref p) => evaluate_pk(p, witness, msg, secp),
+            AstElem::PkV(ref p) => {
+                let res = evaluate_pk(p, witness, msg, secp)?;
+                evaluate_verify(witness)?;
+                Ok(res)
+            }
+            AstElem::PkQ(ref p) => {
+                witness.push(p.to_public_key().to_bytes());
+                Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![]})
+            }
+            AstElem::PkW(ref p) => {
+                let witnesslen = witness.len();
+                witness.swap(witnesslen-1, witnesslen -2);//swap the last elements
+                evaluate_pk(p, witness, msg, secp)
+            }
+            AstElem::Multi(k, ref keys) => evaluate_multi(k, keys, witness, msg, secp),
+            AstElem::MultiV(k, ref keys) => {
+                let res = evaluate_multi(k, keys, witness, msg, secp)?;
+                evaluate_verify(witness)?;
+                Ok(res)
+            }
+            AstElem::TimeT(t) => {
+                let res = evaluate_csv(t, age)?;
+                witness.push(build_scriptint(t as i64));
```

Comment:
I had copy `build_scriptint` from rust bitcoin library as it now a public function and I needed to convert an integer to `Vec<u8>`. This is only for the case where `TimeT` where the return value is not `1`.  

On a similar note, I think all miniscript of `T` form return 1 on success and this is only which is an exception. Is this expression `TimeT` really useful to be added in miniscript?

Reply:
Yeah, unfortunately there are sane usecases where it saves a byte (for example `and_cat(pk(...), time_t(100))` is the shortest way to put a timelock restriction on `pk(...)` in a T context).

We can do a bit simpler than copying `build_scriptint` though .. since timelocks are always nonnegative we can avoid some of the sign craziness.

Reply:
I also wonder whether you can just push `true` onto the stack, or even some poison value, since this element should never be popped by any other Miniscript expression.

Reply:
I think it is safe to push satisfied instead of int `time` for the current semantics of Miniscript, but could not be the case in the future.  What do you think should be done?

Reply:
The `T` calling convention is guaranteed to push a truthy value to the stack. It makes no guarantees about what that value is. If this were to change we'd need a new calling convention for that.

Reply:
So, you suggest that I push the int instead?

Reply:
No, I'm suggesting you push `Satisfied`

Reply:
Yup, that is the current implementation